pandas/io/s3.py

```diff
@@ -1,11 +1,19 @@
-""" s3 support for remote file interactivity """
+from typing import TYPE_CHECKING, Optional, Tuple
 from urllib.parse import urlparse as parse_url
 
 from pandas.compat._optional import import_optional_dependency
 
-s3fs = import_optional_dependency(
-    "s3fs", extra="The s3fs package is required to handle s3 files."
-)
+from pandas._typing import FilePathOrBuffer
+
+""" s3 support for remote file interactivity """
+
+
+if TYPE_CHECKING:
+    import s3fs
+else:
+    s3fs = import_optional_dependency(
```

Comment:
Thinking more about this if you can remove annotations for the optional lib I think would be best. The TYPE_CHECKING block would still require a user to have this installed to validate, so defeats the purpose of optional

Reply:
Ah, good point. I was only thinking about pandas developers running the typechecker.

I will make the return types
```
get_file_and_filesystem(...) -> Tuple[Any, Any]
get_filepath_or_buffer(...) -> Tuple[Any, Optional[str], Optional[str], bool]
```

Reply:
Not super familiar with s3 so if IO works preferable, but otherwise yea makes sense